If async preparing wasm fails, abort() to reject the module promise

stephanreiter opened the pull request:
For example, if a global constructor performs an invalid memory access,
an exception will be thrown that bubbles up to receiveInstantiatedSource.
That causes a rejection of the promise returned from instantiateAsync,
which we should handle.

Fixes #12396

kripken commented:
Thanks for the PR @stephanreiter !

The first CI error appears to be that this succeeds when it should fail:
```python
    err = self.run_js('a.out.js', assert_returncode=NON_ZERO)
```
It does log the error message to the console, but looks like now the return code from the process is 0, which is not good.

kripken commented:
If this is just an issue on node, I wonder if `abort()` could set the return code there, and/or immediately exit with an error. This seems related to the issue with [`NODEJS_CATCH_REJECTION`](https://github.com/emscripten-core/emscripten/blob/1c33ce748393fa2cbc17cbfe1c7a18926ea989cc/src/settings.js#L583) that we don't really have a good answer for yet, but would be great to fix.

kripken commented:
One thought I've had, but am not sure about: in the MODULARIZE case there is a Promise the user can do a `.catch` on. So we don't need to do anything there. But in the non-MODULARIZE case, we do want to turn a Promise rejection into a thrown exception, as there is nothing better to do, and we need to report the error up.

stephanreiter commented:
Maybe like this?
```
#ifdef MODULARIZE
  instantiateAsync().catch(readyPromiseReject);
#else
  instantiateAsync();
#endif
```

stephanreiter commented:
I just tested this for my use case and it does allow me to catch the error. 👍 Let me update the PR to see what CI tests think.

kripken commented:
@stephanreiter Yes, and looks like it passes tests!

As I said I *think* that makes sense, but I'm not sure. The downside is adding a difference in MODULARIZE mode compared to normal mode. But, it does seem like the right thing to me..? If we do it, we'd need to add documentation and testing - in particular I doubt we have a test for MODULARIZE + failing to load the wasm, which we'd need.

But first, it might be good to get more eyes on this, I'll cc some people that may have thoughts or ideas: @curiousdannii @lourd @RReverser (you can start reading from https://github.com/emscripten-core/emscripten/pull/12408#issuecomment-702803243 and/or just read the current code in the PR)

stephanreiter commented:
For MODULARIZE I also believe this to be right. An instantiation failure will reject the module ready promise, and my app code will be informed that something went wrong (and also gets the details via the error object; this is actually better than going through abort() where the error is lost).
Not sure about MODULARIZE=0. I've never used that.
Great to get more eyes on the PR. :-) I'll be happy to invest in tests on Monday.

RReverser commented:
@kripken I think this makes sense too. To be fair, personally I'd prefer never `abort`-ing but instead rejecting / throwing JS issues that can be handled more idiomatically, but that would be a breaking change; meanwhile, adopting this at least for instantiation seems to make sense.

stephanreiter commented:
Thanks for taking a look!

I added a test and rebased my PR.
For the test, I needed to make unhandled rejections in browser tests cause test failures. I'm quite sure this will affect other tests and it will be interesting to see if this uncovers other problems.

stephanreiter commented:
Need to fix how to report unhandled rejections from browser tests.

stephanreiter commented:
We now have two tests of handling of module instantiation failure due to a) a global constructor performing an invalid memory access, and b) download failure. We check that the module promise is rejected and that no promise rejection remains unhandled.

kripken commented on tests/test_browser.py:
```diff
@@ -3361,6 +3366,30 @@ def test_modularize_network_error(self):
+    src = open(path_from_root('tests', 'browser_test_init_crash.cpp')).read()
+    create_test_file('test.cpp', src)
+    browser_reporting_js_path = path_from_root('tests', 'browser_reporting.js')
+    self.compile_btest(['test.cpp', '-s', 'MODULARIZE=1', '-s', 'EXPORT_NAME="createModule"', '--extern-pre-js', browser_reporting_js_path])
```
`compile_btest` can receive the path to the test, avoiding the need to define `src` earlier.

kripken commented on tests/test_browser.py:
```diff
@@ -3361,6 +3366,30 @@ def test_modularize_network_error(self):
+          });
+      </script>
+    ''')
+    self.run_browser('a.html', '...', '/report_result?abort(Runtime error: The application has corrupted its heap memory area (address zero)!)')
```
This error text may change in the future. How about in the C++ throwing a JS exception with specific fixed text? That is,
```cpp
#include <emscripten.h>
...
EM_ASM({
  throw "intentional error to test rejection";
});
```

stephanreiter replied:
Great idea! 😊

kripken reviewed (commented):
Thanks @stephanreiter ! Looks good to me with some test changes, + also please mention this in the Changelog.

lourd commented:
Good changes, thanks @stephanreiter! LGTM.

This looks similar to my PR #11625 that got reviewed and then fell along the wayside. @kripken, let me know if you'd like to see that one finished. I think the async control flow between instantiation, fallback instantiation, and receiving the instantiated module is still confusing.

lourd commented on tests/test_browser.py:
```diff
@@ -3348,6 +3348,11 @@ def test_modularize_network_error(self):
     create_test_file('a.html', '''
       <script src="a.out.js"></script>
       <script>
+        if (typeof window === 'object' && window) {
```
Is the `&& window` needed?

stephanreiter replied:
I also think it is not needed. window is either valid or undefined. And typeof undefined is 'undefined'.
(typeof null is 'object', which is ... special ... but doesn't apply here as window will never be null).

lourd commented on tests/test_browser.py:
```diff
@@ -3361,6 +3366,30 @@ def test_modularize_network_error(self):
+    create_test_file('a.html', '''
+      <script src="a.out.js"></script>
+      <script>
+        if (typeof window === 'object' && window) {
```
Same here, I don't believe the extra `&& window` is necessary

kripken commented:
>  @kripken, let me know if you'd like to see that one finished. 

I think refactoring that code would be great, sure!

kripken commented on tests/browser_test_init_throws.cpp:
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright 2015 The Emscripten Authors.  All rights reserved.
```
```suggestion
 * Copyright 2020 The Emscripten Authors.  All rights reserved.
```

stephanreiter replied:
Thanks!

kripken commented on tests/test_browser.py:
```diff
@@ -3347,13 +3347,17 @@ def test_modularize(self):
+          window.addEventListener('unhandledrejection', function(event) {
+            reportResultToServer("Unhandled promise rejection: " + event.reason.message);
+          });
+        }
```
Are the changes to `test_modularize_network_error` just cleanups, and to emit a nicer error, or are they necessary for some reason?

stephanreiter replied:
I kept the clean up bits and removed the listening to unhandledrejection again.
Ideally, all browser tests should fail on unhandledrejection. But that is better addressed in a separate commit.

sbc100 commented on tests/test_browser.py:
```diff
@@ -3367,6 +3371,29 @@ def test_modularize_network_error(self):
     self.run_browser('a.html', '...', '/report_result?abort(both async and sync fetching of the wasm failed)')
 
+  def test_modularize_init_error(self):
+    test_cpp_path = path_from_root('tests', 'browser_test_init_throws.cpp')
```
Can you put this in `tests/browser` ad call it `test_modularize_init_error.cpp` to match the test name?

stephanreiter replied:
Done!

kripken reviewed (approved):
Great!
